src/bsd.rs

```diff
@@ -0,0 +1,325 @@
+const RTM_ADDRS: i32 = libc::RTA_DST | libc::RTA_IFP;
+
+#[cfg(apple)]
+const ALIGN: usize = size_of::<libc::c_int>();
+
+#[cfg(any(target_os = "freebsd", target_os = "netbsd", target_os = "openbsd"))]
+// See https://github.com/freebsd/freebsd-src/blob/524a425d30fce3d5e47614db796046830b1f6a83/sys/net/route.h#L362-L371
+// See https://github.com/NetBSD/src/blob/4b50954e98313db58d189dd87b4541929efccb09/sys/net/route.h#L329-L331
+const ALIGN: usize = size_of::<libc::c_long>();
+
+use crate::{aligned_by, default_err};
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const AF_INET_U8: u8 = AF_INET as u8;
+const_assert_eq!(AF_INET_U8 as i32, AF_INET);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const AF_INET6_U8: u8 = AF_INET6 as u8;
+const_assert_eq!(AF_INET6_U8 as i32, AF_INET6);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const AF_LINK_U8: u8 = AF_LINK as u8;
+const_assert_eq!(AF_LINK_U8 as i32, AF_LINK);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const RTM_VERSION_U8: u8 = RTM_VERSION as u8;
+const_assert_eq!(RTM_VERSION_U8 as i32, RTM_VERSION);
+
+#[allow(clippy::cast_possible_truncation)] // Guarded by the following `const_assert_eq!`.
+const RTM_GET_U8: u8 = RTM_GET as u8;
+const_assert_eq!(RTM_GET_U8 as i32, RTM_GET);
+
+const_assert!(size_of::<sockaddr_in>() + ALIGN <= u8::MAX as usize);
+const_assert!(size_of::<sockaddr_in6>() + ALIGN <= u8::MAX as usize);
+const_assert!(size_of::<rt_msghdr>() <= u8::MAX as usize);
+
+struct IfAddrs(*mut ifaddrs);
+
+impl Default for IfAddrs {
+    fn default() -> Self {
+        Self(ptr::null_mut())
+    }
+}
+
+impl IfAddrs {
+    fn new() -> Result<Self> {
+        let mut ifap = Self::default();
+        // getifaddrs allocates memory for the linked list of interfaces that is freed by
+        // `IfAddrs::drop`.
+        if unsafe { getifaddrs(ptr::from_mut(&mut ifap.0)) } != 0 {
+            return Err(Error::last_os_error());
+        }
+        Ok(ifap)
+    }
+
+    const fn iter(&self) -> IfAddrPtr {
+        IfAddrPtr {
+            ptr: self.0,
+            _ref: PhantomData,
+        }
+    }
+}
+
+impl Drop for IfAddrs {
+    fn drop(&mut self) {
+        if !self.0.is_null() {
+            // Free the memory allocated by `getifaddrs`.
+            unsafe { freeifaddrs(self.0) };
+        }
+    }
+}
+
+struct IfAddrPtr<'a> {
+    ptr: *mut ifaddrs,
+    _ref: PhantomData<&'a ifaddrs>,
+}
+
+impl IfAddrPtr<'_> {
+    fn addr(&self) -> libc::sockaddr {
+        unsafe { *self.ifa_addr }
+    }
+
+    fn name(&self) -> String {
+        unsafe { CStr::from_ptr(self.ifa_name).to_string_lossy().to_string() }
+    }
+
+    fn data(&self) -> Option<if_data> {
+        if self.ifa_data.is_null() {
+            None
+        } else {
+            Some(unsafe { *(self.ifa_data as *const if_data) })
```

Comment:
```suggestion
            Some(unsafe { self.ifa_data.read() })
```

I think that you can just read rather than casting then dereferencing.  If you need a cast, then use `.cast()`.